exercises/concept/assembly-line/bats-jq.bash

```diff
@@ -0,0 +1,29 @@
+#!/usr/bin/env bash
+#
+# `bats-core` will consume both stdout and stderr for the `run` command's output.
+# However `jq` prints its DEBUG output on stderr.
+#
+# Lines starting with `["DEBUG:",` will be prefixed with a hash and printed on file descriptor 3.
+# Other lines on stderr will remain on stderr for bats to consume.
+#
+# See `bats-core` docs:
+# - "Printing to the terminal", https://bats-core.readthedocs.io/en/stable/writing-tests.html#printing-to-the-terminal
+# - "File descriptor 3", https://bats-core.readthedocs.io/en/stable/writing-tests.html#file-descriptor-3-read-this-if-bats-hangs
+
+
+jq() {
+    local output stderr rc
+    stderr=$(mktemp)
```

Comment:
Would it be worth using a `trap` for cleanup?

Reply:
I'm not sure how many times this function will be called per shell. A quick experiment shows that traps do not accumulate
```
$ bash -c 'trap "echo first" EXIT; trap "echo second" EXIT'
second
```

Reply:
Yeah. Traps are signal handlers. You can only install one handler per signal. But that's fine. You can use a global for the tempfile and have the EXIT handler delete the tempfile if it exists. That keeps things clean.